paymentgateway/paypal/classes/paymentgateway.php

```diff
@@ -0,0 +1,197 @@
+        $eventdata->userto            = $admin;
+        $eventdata->subject           = "PAYPAL ERROR: ".$subject;
+        $eventdata->fullmessage       = $message;
+        $eventdata->fullmessageformat = FORMAT_PLAIN;
+        $eventdata->fullmessagehtml   = '';
+        $eventdata->smallmessage      = '';
+        message_send($eventdata);
+    }
+
+    public function submit_purchase_data($data) {
+        global $DB;
+        $status = $data->payment_status;
+
+        $paymentstatus = 0;
+        if ($status == "Completed" || $status == "Processed") {
+            $paymentstatus = 1;
+        } else if ($status == "Pending") {
+            $paymentstatus = 2;
+        }
+
+        // Check for duplicate txn ids
+        $record = $DB->get_record('paymentgateway_paypal', ['txn_id' => $data->txn_id]);
+        if ($record != false) {
+            $paymentstatus = 0;
+        }
+
+        $res = \tool_paymentplugin\payment_manager::submit_transaction($paymentstatus, 'paymentgateway_paypal', $this->name, $data->userid,
+            $data->mc_currency, $data->mc_gross, $data->payment_date, $data->courseid, $data);
+
+        if ($res == 0) { // ERROR
+            $this->message_paypal_error_to_admin("Invalid Payment.", $data);
+            return 0;
+        } else if ($res == 2) { // PENDING
+            $this->message_paypal_error_to_admin("Payment Pending.", $data);
+            return 2;
+        }
+        return 1; // SUCCESS
+    }
+
+    public function payment_button($courseid) {
+        global $CFG, $USER, $DB;
+
+        // Gather config data.
+        $sandboxid       = get_config('paymentgateway_paypal', 'clientidsandbox');
+        $productionid    = get_config('paymentgateway_paypal', 'clientidproduction');
+        // If testing, set '$CFG->usepaypalsandbox = 1;' in config.php.
+        $clientid        = empty($CFG->usepaypalsandbox) ? $productionid : $sandboxid;
+        $buttonsize      = 'small';
+        $buttoncolour    = get_config('paymentgateway_paypal', 'colour');
+        $buttonshape     = get_config('paymentgateway_paypal', 'shape');
+        $course          = $DB->get_record('course', array('id' => $courseid));
+        $context         = \context_course::instance($course->id);
+        $coursefullname  = format_string($course->fullname, true, array('context' => $context));
+        $userfirstname   = $USER->firstname;
+        $userlastname    = $USER->lastname;
+        $useremail       = $USER->email;
+        $record          = $DB->get_record('tool_paymentplugin_course', ['courseid' => $course->id]);
+        $cost            = $record->cost;
+        $amount          = number_format((float)$cost, 2, '.', '');
+        $currency        = get_config('tool_paymentplugin', 'currency');
+        // Custom parameter that holds user ID and course ID for the IPN page to read.
+        $custom          = $USER->id . '-' . $course->id;
+
+        // Gateway HTML.
+        $html = <<<HTML
```

Comment:
This ended up being quite a large chunk of HTML + JS. I would like the JS to be moved into an AMD module, and the HTML to be escaped using html_writer